core/controllers/reader.py

```diff
@@ -1274,3 +1299,103 @@ def put(self, entity_type, entity_id):
             entity_type, state_reference,
             interaction_id, answer, answer_details)
         self.render_json({})
+
+
+class CheckpointCompletedEventHandler(base.BaseHandler):
+    """Tracks a learner completing a checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'last_completed_checkpoint_exp_version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'last_completed_checkpoint_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                }
+            },
+        }
+    }
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    @acl_decorators.can_play_exploration
+    def put(self, exploration_id):
+        """Handles PUT requests.
+
+        Args:
+            exploration_id: str. The ID of the exploration.
+        """
+
+        user_id = self.user_id
+        last_completed_checkpoint_state_name = self.normalized_payload.get(
+            'last_completed_checkpoint_state_name')
+        last_completed_checkpoint_exp_version = self.normalized_payload.get(
+            'last_completed_checkpoint_exp_version')
+
+        user_services.update_learner_checkpoint_progress(
+            user_id,
+            exploration_id,
+            last_completed_checkpoint_state_name,
+            last_completed_checkpoint_exp_version)
+
+        self.render_json(self.values)
+
+
+class CheckpointVisitedEventHandler(base.BaseHandler):
+    """Tracks a learner visiting a previously completed checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'last_completed_checkpoint_exp_version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'latest_completed_checkpoint_state_name': {
```

Comment:
Should it not be 'latest_visited_checkpoint_state_name' ?

Reply:
Done!